router.go

```diff
@@ -67,51 +85,123 @@ func (m MapRouter) Register(rs []transport.Procedure) {
+			panic(fmt.Sprintf("Cannot register a handler for both (service, procedure) on any * encoding and (service, procedure, encoding), specifically (%q, %q, %q)", r.Service, r.Name, r.Encoding))
+		}
+		// Route to individual handlers for unique combinations of service,
+		// procedure, and encoding. This shall henceforth be the
+		// recommended way for models to register procedures.
+		m.serviceProcedureEncodings[spe] = r
+		// Record supported encodings.
+		m.supportedEncodings[sp] = append(m.supportedEncodings[sp], string(r.Encoding))
 	}
 }
 
 // Procedures returns a list procedures that
 // have been registered so far.
 func (m MapRouter) Procedures() []transport.Procedure {
-	procs := make([]transport.Procedure, 0, len(m.entries))
-	for _, v := range m.entries {
+	procs := make([]transport.Procedure, 0, len(m.serviceProcedures)+len(m.serviceProcedureEncodings))
+	for _, v := range m.serviceProcedures {
+		procs = append(procs, v)
+	}
+	for _, v := range m.serviceProcedureEncodings {
 		procs = append(procs, v)
 	}
-	sort.Sort(proceduresByServiceProcedure(procs))
+	sort.Sort(sortableProcedures(procs))
 	return procs
 }
 
-// Choose retrives the HandlerSpec for the service and procedure noted on the
-// transport request, or returns an unrecognized procedure error (testable with
-// transport.IsUnrecognizedProcedureError(err)).
+type sortableProcedures []transport.Procedure
+
+func (ps sortableProcedures) Len() int {
+	return len(ps)
+}
+
+func (ps sortableProcedures) Less(i int, j int) bool {
+	return ps[i].Less(ps[j])
+}
+
+func (ps sortableProcedures) Swap(i int, j int) {
+	ps[i], ps[j] = ps[j], ps[i]
+}
+
+// Choose retrives the HandlerSpec for the service, procedure, and encoding
+// noted on the transport request, or returns an unrecognized procedure error
+// (testable with transport.IsUnrecognizedProcedureError(err)).
 func (m MapRouter) Choose(ctx context.Context, req *transport.Request) (transport.HandlerSpec, error) {
-	service, procedure := req.Service, req.Procedure
+	service, procedure, encoding := req.Service, req.Procedure, req.Encoding
 	if service == "" {
 		service = m.defaultService
 	}
 
-	sp := serviceProcedure{
+	// Fully specified combinations of service, procedure, and encoding.
+	spe := serviceProcedureEncoding{
 		service:   service,
 		procedure: procedure,
+		encoding:  encoding,
 	}
-	if procedure, ok := m.entries[sp]; ok {
+	if procedure, ok := m.serviceProcedureEncodings[spe]; ok {
 		return procedure.HandlerSpec, nil
 	}
 
-	return transport.HandlerSpec{}, transport.UnrecognizedProcedureError(req)
-}
+	// Alternately use the original behavior: route all encodings to the same
+	// handler.
+	sp := serviceProcedure{
+		service:   service,
+		procedure: procedure,
+	}
+	if procedure, ok := m.serviceProcedures[sp]; ok {
+		return procedure.HandlerSpec, nil
+	}
 
-type proceduresByServiceProcedure []transport.Procedure
+	// Supported procedure, unrecognized encoding.
+	if want, ok := m.supportedEncodings[sp]; ok {
```

Comment:
Because code is read more than it is written.  My philosophy is to bring up anything that slowed down my understanding of the code if I can offer an alternative.

In this particular case, I jumped to the `len(want) == 1` and scratched my head a bit because we start treating this want as a list, which is not insinuated at all by the variables naming, which is confusing.